Cycles various lights update

boberfly opened the pull request:
This is a precursor for an upcoming PR update to support USD Lux lights and USDPreviewSurface and friends.

Generally describe what this PR will do, and why it is needed

- Added `use_camera` and `use_caustics` light parameters
- Use degrees for the `distant_light` angle parameter
- Replace the uniform `size` with `sizeu` and `sizev` for disk, quad and portals
- Fixed a bug where quad lights become round by setting this explicitly to false
- Added virtual `normalize` parameter which multiplies based on the light's surface area (it will be inaccurate if scale is used on the light)

### Related issues ###

- NA

### Dependencies ###

- NA

### Breaking changes ###

- Breaks existing quad/disk/portals which use the size parameter.
- Breaks distant light angle that implied radians and not degrees.

### Checklist ###

- [x] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [ ] I have updated the documentation, if applicable.
- [ ] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [x] My code follows the Gaffer project's prevailing coding style and conventions.

danieldresser-ie commented:
John asked me to take a look at the normalization math.  The quad, disk, and sphere types all seem quite uncontroversial, since they have a defined area, and match to https://developer.blender.org/diffusion/C/browse/master/src/hydra/light.cpp

I'm not exactly sure what "strength" would mean on a portal, and whether a portal would be normalized ( I haven't played with Cycles portals ).  I would have thought that the brightness and spread of a portal would be defined by what is shining through it, and normalization wouldn't really make sense?

The distant light seems like the one that could be a bit odd ... and this is the case where Blender's hydra delegate does not appear to apply unnormalization.  Artists who are used to packages where lights aren't normalized, probably still don't expect distant lights to be unnormalized - from my perspective, increasing the angle of a distant light is usually conceptualized as increasing the range of angles coming out of an emitter that is infinitely large anyway, rather than making something larger that somehow exists in angular space.  Setting up a hard sunlight with 0 softness and having the intensity set to 0 is probably not usually what would be expected.  Perhaps it would make sense to follow the lead of the hydra delegate and not have a normalization control on distant lights?

boberfly commented:
Hey @danieldresser-ie 

I was actually the one who made the original pass on (un)normalized lights for the HDCycles/Blackbird delegate and did some extensive testing here:
https://github.com/tangent-opensource/hdBlackbird/pull/144

I did add some fudge factor values by increasing exposure to match the other delegates which I'm considering for the UsdLux conversion pass.

The angle stuff did match one of the other delegates but I don't remember now, it might've been PRMan and/or Karma. Yeah agreed an un-normalized directional light with 0 angle could be confusing for artists when the light goes out.

Some digging in Cycles code makes me think the most trivial way could be to just add the normalize as a socket and make this value be set to 1.0f when the socket is false, but I've not tried yet:
https://github.com/blender/cycles/blob/master/src/scene/light.cpp#L867

Anyways cheers for the review!

danieldresser-ie commented:
I should have looked a bit more into this - you're totally right that these other renderers have a normalized distant light that we should match.  The specific math doesn't look quite right though ... just squaring the angle isn't going to behave right for large angles.  Empirical testing with Arnold appears to show that Arnold normalizes based on the solid angle of the distant light, ie:
`normalized = unnormalized / ( 2 * pi * ( 1 - cos( 0.5 * angle * pi / 180 ) ) )`, where `angle` is the full apex angle of the light cone ( I tend to think about it terms of the max deviation angle from forward, which is half that ).  That's probably the factor you should be multiplying with?

boberfly commented:
Yeah in Cycles it is this:
https://github.com/blender/cycles/blob/master/src/scene/light.cpp#L798
```
float angle = light->angle / 2.0f;
float radius = tanf(angle);
float area = M_PI_F * radius * radius;
float invarea = (area > 0.0f) ? 1.0f / area : 1.0f;
```
or in the `constantLightStrength` function:
```
const float angle = IECore::degreesToRadians( parameterValue<float>( lightShader->parameters(), "angle", 0.0f ) ) / 2.0f;
if( angle > 0.0f )
{
    const float radius = tanf(angle);
    strength *= M_PI_F * radius * radius;
}
```
I'm amazed how you figured it out from Arnold there, do you have source access or just a solid understanding and matched it? :)

Cheers I'll update with the Cycles one I think, if we plan to gut this code out later if my patch back to Cycles is successful... I remember one of those renderers I tested (might've been Arnold) was the one that was a bit different.

danieldresser-ie commented:
That was just evaluating a set of angles in Arnold, and finding which equation matched.  I think there's a strong case to be made that Arnold's approach is objectively correct anyway - the coverage of the distant light is an angular value, and should be measured in steradians.  Blenders radius should match as long as angle is small, but looks like it explodes as the angle approaches 180.  But yeah, for this, it's right to match Blender's code.

murraystevenson commented on src/GafferCycles/SocketHandler.cpp:
```diff
@@ -514,27 +514,29 @@ void setupLightPlugs( const std::string &shaderName, const ccl::NodeType *nodeTy
-	if( shaderName == "spot_light" )
+	if( shaderName == "portal" || shaderName == "quad_light" || shaderName == "disk_light" )
+	{
+		validPlugs.insert( setupTypedPlug<FloatPlug>( "sizeu", plugsParent, Gaffer::Plug::In, 2.0f ) );
+		validPlugs.insert( setupTypedPlug<FloatPlug>( "sizev", plugsParent, Gaffer::Plug::In, 2.0f ) );
```
What do you think about representing these as `width` and `height` rather than `sizeu` and `sizev`? It's true we often present a renderer's parameters mostly raw and unvarnished to the user, but `width` and `height` are likely more familiar/expected names as well as being closer to UsdLuxLight.

We'll likely add the same to Arnold quad lights at some point in the near future too.

johnhaddon replied:
I'm inclined to agree with Murray here - what do you think @boberfly? Are you likely to be able to find time to get this over the line, or would you like us to make those final few tweaks?

boberfly replied:
Hey @murraystevenson & @johnhaddon apologies this wasn't addressed before the release, but I got this sorted now with width/height! This reads better than sizeu/sizev.

murraystevenson commented on startup/GafferScene/cyclesLights.py:
```diff
@@ -61,14 +61,19 @@
 Gaffer.Metadata.registerValue( "cycles:light:disk_light", "colorParameter", "color" )
-Gaffer.Metadata.registerValue( "cycles:light:disk_light", "widthParameter", "size" )
+Gaffer.Metadata.registerValue( "cycles:light:disk_light", "widthParameter", "sizeu" )
+Gaffer.Metadata.registerValue( "cycles:light:quad_light", "heightParameter", "sizev" )
```
There's duplication here of the quad_light `heightParameter` registration, though I think you'll find that if you change this to `cycles:light:disk_light` it'll have no effect as the disk light [visualiser](https://github.com/GafferHQ/gaffer/blob/c618bef83170b5c5a5221cc03f95870ba03a99cb/src/GafferSceneUI/StandardLightVisualiser.cpp#L562-L563) only supports `widthParameter` or `radiusParameter`. 

Do you think it's worth keeping the `disk_light` limited to width for now rather than presenting a potentially confusing visualisation? UsdLuxDiskLight also only supports radius, so I don't think we'd be losing anything when it comes to any potential future USD conversions that may be cooking. :)

boberfly replied:
Oops copy-pasta typo. Yep I've got it set now to just apply width to both sizeu & sizev now.

murraystevenson reviewed (changes requested):
Cheers for this Alex! It looks like quite a nice improvement to the lights. I've one or two comments/questions inline mostly about naming and how we present the lights to users. I'd be interested in hearing your thoughts.

When making breaking changes like this, it can be helpful to provide a compatibility [config](https://github.com/GafferHQ/gaffer/blob/main/startup/Gaffer/randomCompatibility.py) remapping the plugs in scripts saved in earlier versions of Gaffer. We have a bit of leeway to break things while Cycles is still under feature preview status, but it's worth keeping in mind for future changes.

It'd also be good to update Changes.md with a description of these changes please. :)

boberfly commented:
Alright I've addressed the notes that you guys found. I've also got a patch for normalised lights that @brechtvl has agreed to upstream into Cycles eventually, and that is [here](https://developer.blender.org/D16838).  It *should* just work as a patch in the dependencies source if we want it to live there before we wait for Cycles 3.5(?) instead of having the un-normalise code in this PR.

I've also tweaked the distant light un-normalise calculation to match what it'll do when the upstream patch gets applied.

> It'd also be good to update Changes.md with a description of these changes please. :)

And I've updated Changes.md - I'm always hesitant to do this as it feels like I'm imposing the next version number, should I just go 1.x.x.x ?

boberfly commented:
Looks like it got merged:
https://developer.blender.org/D16838

3.5.0 isn't out, but should we scrap the extra client-side code in Gaffer and just hold this as a  patch for dependencies?

johnhaddon commented:
> Looks like it got merged:

Nice one!

> 3.5.0 isn't out, but should we scrap the extra client-side code in Gaffer and just hold this as a patch for dependencies?

I don't want to get into the habit of cherry-picking future Cycles features into GafferHQ/dependencies - ideally patches are just limited to temporary build fixes. Since you've already done the work for the client-side code in this PR, and it's already reviewed, I think we can just merge it and then remove the manual normalisation when we update to 3.5.

@murraystevenson, as you reviewed before, could you give a quick look to the latest changes with a view to merging please?

murraystevenson commented on Changes.md:
```diff
@@ -1,3 +1,15 @@
+
+- Cycles :
+  - Added use_camera and use_caustics for lights.
+  - Fixed distant light angle is in degrees and not radians.
```
To be entirely pedantic, this should probably go under a `Fixes` header.

boberfly replied:
Done

murraystevenson reviewed (commented):
Thanks Alex! Those adjustments feel a lot better to me. Though can we go ahead and hide the nodules that have cropped up for `width` and `height`? They clutter the UI and give the impression that those parameters are connectable.

> And I've updated Changes.md - I'm always hesitant to do this as it feels like I'm imposing the next version number, should I just go 1.x.x.x ?

Go with `1.1.x.x` like you've done in #5044, as this PR is targeting `1.1_maintenance`. The remaining version numbers will be specified at the time of release so you don't have to worry about those. For background, there's a bit of info on that release procedure in the [Wiki](https://github.com/GafferHQ/gaffer/wiki/Making-a-Release).

> Looks like it got merged:

Nice work! :)

boberfly commented:
> Thanks Alex! Those adjustments feel a lot better to me. Though can we go ahead and hide the nodules that have cropped up for `width` and `height`? They clutter the UI and give the impression that those parameters are connectable.
> 

Ahh nice find, fixed.

Rebased on 1.1 maintenance, just waiting on the CI but I think we're good now to merge.

murraystevenson reviewed (approved):
> Rebased on 1.1 maintenance, just waiting on the CI but I think we're good now to merge.

Nice one, LGTM. Cheers mate!
